larpandoracontent/LArHelpers/LArDiscreteProbabilityHelper.cc

```diff
@@ -0,0 +1,129 @@
+/**
+ *  @file   larpandoracontent/LArHelpers/LArDiscreteProbabilityHelper.cc
+ *
+ *  @brief  Implementation of the discrete probability helper helper class.
+ *
+ *  $Log: $
+ */
+
+#include "larpandoracontent/LArHelpers/LArDiscreteProbabilityHelper.h"
+
+namespace lar_content
+{
+
+template <typename T>
+float LArDiscreteProbabilityHelper::CalculateCorrelationCoefficientPValueFromPermutationTest(const T &t1, const T &t2, 
+    std::mt19937 &randomNumberGenerator, const size_t nPermutations)
+{
+    if (1 > nPermutations)
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+    const float rNominal(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficient(t1,t2));
+
+    int nExtreme(0);
+    for (unsigned int iPermutation = 0; iPermutation < nPermutations; ++iPermutation)
+    {
+        const float rRandomised(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficient(
+            LArDiscreteProbabilityHelper::MakeRandomisedSample(t1,randomNumberGenerator),
+            LArDiscreteProbabilityHelper::MakeRandomisedSample(t2,randomNumberGenerator)));
+
+        if ((rRandomised-rNominal) > std::numeric_limits<float>::epsilon())
+            nExtreme++;
+    }
+
+    return static_cast<float>(nExtreme)/static_cast<float>(nPermutations);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename T>
+float LArDiscreteProbabilityHelper::CalculateCorrelationCoefficientPValueFromStudentTDistribution(const T &t1, 
+    const T &t2, const size_t nIntegrationSteps)
+{
+    const float correlation(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficient(t1,t2));
+    const float dof(static_cast<float>(LArDiscreteProbabilityHelper::GetSize(t1)) - 2.f);
+    const float tTestStatistic(correlation*sqrt(dof)/(sqrt(1.f - correlation*correlation)));
```

Comment:
The 1s and 2s are exact and won't change (they are part of the maths for the t-dist and number of degrees of freedom.
The 15 in principle can change as it defines how far along the distribution to integrate.  The 15 was chosen so that you're really really far along the distribution.  I can make the 15 a function argument.  This makes the most sense, given that the number of steps is a function argument.

Reply:
Done